lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp

```diff
@@ -2235,6 +2235,143 @@ static Constant *getConstant(const NodeValue &NV) {
   return dyn_cast<Constant>(N);
 }
 
+static bool foldBatchNormalizeIntoConv(Module *M, BatchNormalizationNode *BN,
+                                       ConvolutionNode *CV) {
+  bool normalizationHappened = false;
+  switch (CV->getElementType(ConvolutionNode::ResultIdx)) {
+  case ElemKind::FloatTy:
+    normalizationHappened = normalizeWeights<float>(M, *CV, *BN);
+    break;
+  case ElemKind::Float16Ty:
+    normalizationHappened = normalizeWeights<float16_t>(M, *CV, *BN);
+    break;
+  case ElemKind::BFloat16Ty:
+    normalizationHappened = normalizeWeights<bfloat16_t>(M, *CV, *BN);
+    break;
+  default:
+    llvm_unreachable("Type not supported");
+  }
+
+  if (!normalizationHappened) {
+    return normalizationHappened;
+  }
+
+  // Take the predicate of what was expected for the output.
+  CV->setPredicate(BN->getPredicate());
+  BN->getResult().replaceAllUsesOfWith(CV);
+  return normalizationHappened;
+}
+
+// Normalize the weight of \p FullyConnectedNode with what \p BN is doing, given
+// containing
+/// Module \p M. \returns whether or not the normalization was possible.
+template <typename ElemTy>
+bool normalizeFullyConnectedWeights(Module *M, FullyConnectedNode &FC,
```

Comment:
This function is almost an exact copy of `normalizeWeights`. Everything is identical except for operand access code and some variable names. I think it's possible to have a single function which would cover both Conv and FC. We can pass to the function the node as `Node *` and indices (`ConvolutionNode::FilterIdx`/`ConvolutionNode::BiasIdx` and `FullyConnectedNode::WeightsIdx`/`FullyConnectedNode::BiasIdx`) to be able to access its operands. This is should cover all kind-specific uses of the node in this function.